datafusion/src/logical_plan/builder.rs

```diff
@@ -1395,4 +1508,58 @@ mod tests {
         assert!(stringified_plan.should_display(true));
         assert!(!stringified_plan.should_display(false));
     }
+
+    #[test]
+    fn invalid_plan_with_not_non_boolean_expression() -> Result<()> {
+        let mut plan = LogicalPlanBuilder::scan_empty(
+            Some("employee_csv"),
+            &employee_schema(),
+            Some(vec![0, 1, 2, 3, 4]),
+        )?;
+        plan = plan.project(vec![Expr::Not(Box::new(Expr::Column(Column {
+            relation: None,
+            name: "id".to_string(),
+        })))])?;
+        match plan.build() {
+            Ok(_) => panic!("Unexpect Ok"),
+            Err(e) => {
+                assert_eq!(
+                    e.to_string(),
+                    "Error during planning: \
+                    Invalid Not Expression(NOT #employee_csv.id) -- \
+                    Type(Int32) of expression(#employee_csv.id) must be of a boolean type"
+                )
+            }
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn invalid_plan_cast_utf8_boolean() -> Result<()> {
+        let mut plan = LogicalPlanBuilder::scan_empty(
+            Some("employee_csv"),
+            &employee_schema(),
+            Some(vec![0, 1, 2, 3, 4]),
+        )?;
+        let col = Expr::Column(Column {
+            relation: None,
+            name: "first_name".to_owned(),
+        });
+        plan = plan.project(vec![Expr::Cast {
+            expr: Box::new(col),
+            data_type: DataType::Boolean,
+        }])?;
+        match plan.build() {
+            Ok(_) => panic!("Unexpect Ok"),
+            Err(e) => {
+                assert_eq!(
+                    e.to_string(),
+                    "Error during planning: \
+                    Invalid Cast Expression(CAST(#employee_csv.first_name AS Boolean)) -- \
+                    Utf8 cannot cast to Boolean"
+                )
+            }
+        }
+        Ok(())
+    }
```

Comment:
Can we also please add a test for using wildcards incorrectly -- like `SELECT .. GROUP BY *` and  `SELECT .. GROUP BY foo*`?